crates/hir/src/lib.rs

```diff
@@ -1483,9 +1484,49 @@ impl Impl {
 
         inherent.all_impls().chain(trait_.all_impls()).map(Self::from).collect()
     }
-    pub fn for_trait(db: &dyn HirDatabase, krate: Crate, trait_: Trait) -> Vec<Impl> {
-        let impls = db.trait_impls_in_crate(krate.id);
-        impls.for_trait(trait_.id).map(Self::from).collect()
+
+    pub fn all_for_type(db: &dyn HirDatabase, Type { krate, ty }: Type) -> Vec<Impl> {
+        let def_crates = match ty.value.def_crates(db, krate) {
+            Some(def_crates) => def_crates,
+            None => return vec![],
+        };
+
+        let filter = |impl_def: &Impl| {
+            let target_ty = impl_def.target_ty(db);
+            let rref = target_ty.remove_ref();
+            ty.value.equals_ctor(rref.as_ref().map_or(&target_ty.ty.value, |it| &it.ty.value))
+        };
+
+        let mut all = Vec::new();
+        def_crates.iter().for_each(|&id| {
+            all.extend(db.inherent_impls_in_crate(id).all_impls().map(Self::from).filter(filter))
+        });
+        let fp = TyFingerprint::for_impl(&ty.value);
+        for id in def_crates
+            .iter()
+            .flat_map(|&id| Crate { id }.reverse_dependencies(db))
+            .map(|Crate { id }| id)
```

Comment:
So this is interesting. The thing is, *downstream* crates also can add impls of their traits to the type, so we probably want to look at the crate graph here.

Reply:
Haha, I knew I forgot something but I couldn't think of what 😄

Reply:
But... this was already looking at the reverse dependencies of the definition crates, wasn't it?

Reply:
Actually ye it did

Reply:
But those aren't transitive are they? So it was still wrong I think?

Reply:
Ah, hm, well. Only direct reverse dependencies can add impls, _unless_ the type gets reexported...

Reply:
I wonder whether the other use of `reverse_dependencies` doesn't have the same problem though :thinking:

Reply:
I think so, made an implementation for it in #8034